lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp

```diff
@@ -889,4 +981,182 @@ int handle_sm_message_routing_failure(
+      cause;
+  container_len++;
+
+  dlmsg->payload_container.len = container_len;
+  len += PAYLOAD_CONTAINER_LENGTH;
+  len += dlmsg->payload_container.len;
+
+  /* Ciphering algorithms, EEA1 and EEA2 expects length to be mode of 4,
+   * so length is modified such that it will be mode of 4
+   */
+  AMF_GET_BYTE_ALIGNED_LENGTH(len);
+  if (SECURITY_HEADER_TYPE_NOT_PROTECTED != msg->header.security_header_type) {
+    amf_msg_header* header = &msg->security_protected.plain.amf.header;
+    /*
+     * Expand size of protected NAS message
+     */
+    len += NAS_MESSAGE_SECURITY_HEADER_SIZE;
+    /*
+     * Set header of plain NAS message
+     */
+    header->extended_protocol_discriminator = M5G_MOBILITY_MANAGEMENT_MESSAGES;
+    header->security_header_type = SECURITY_HEADER_TYPE_NOT_PROTECTED;
+  }
+
+  return len;
+}
+
+/****************************************************************************
+ **                                                                        **
+ ** Name        :  amf_pdu_session_establishment_reject()                  **
+ **                                                                        **
+ ** Description :  Send the Downlink Transport with 5GMM Cause to gnb      **
+ **                                                                        **
+ ** Inputs      :  amf_ue_ngap_id_t : pdusession response message          **
+ **                session_id	    : PDU Session Inputs                   **
+ **		   pti              : Procedure transaction identity       **
+ **		   5GSM cause       : 5GSM cause                           **
+ **                                                                        **
+ **  Return     :  RETURNok, RETURNerror                                   **
+ **                                                                        **
+ ***************************************************************************/
+int amf_pdu_session_establishment_reject(
+    amf_ue_ngap_id_t ue_id, uint8_t session_id, uint8_t pti, uint8_t cause) {
+  nas5g_error_code_t rc = M5G_AS_SUCCESS;
+  uint32_t bytes        = 0;
+  uint32_t len          = 0;
+  bstring buffer;
+  ue_m5gmm_context_s* ue_context = nullptr;
+  amf_nas_message_t msg          = {};
+
+  ue_context = amf_ue_context_exists_amf_ue_ngap_id(ue_id);
+
+  if (!ue_context) {
+    OAILOG_ERROR(
+        LOG_AMF_APP, "UE Context not found for UE ID: " AMF_UE_NGAP_ID_FMT,
+        ue_id);
+    return M5G_AS_FAILURE;
+  }
+
+  len = construct_pdu_session_reject_dl_req(
+      ue_context->amf_context._security.dl_count.seq_num, session_id, pti,
+      cause, true, &msg);
+
+  if (len <= 0) {
+    OAILOG_WARNING(LOG_AMF_APP, "PDU Construction is failed \n");
+    return M5G_AS_FAILURE;
+  }
+  buffer = bfromcstralloc(len, "\0");
+  bytes  = nas5g_message_encode(
+      buffer->data, &msg, len, &ue_context->amf_context._security);
+
+  if (bytes > 0) {
+    buffer->slen = bytes;
+    amf_app_handle_nas_dl_req(ue_id, buffer, rc);
```

Comment:
Is the memory allocated for `buffer` always released inside `amf_app_handle_nas_dl_req`?

Reply:
Yes, Memory allocated for buffer always released inside amf_app_handle_nas_dl_req.
![buffer destroy](https://user-images.githubusercontent.com/89978170/141515964-b578471b-55f9-472f-ac95-4ea500391c18.JPG)

Reply:
@sreedharkumartn , note that the bstring nas_msg is not destroyed if the function returns due to null value of amf_app_desc_p or null ue_context. Can you create a separate PR to destroy nas_msg in those cases as well?